WebGPU 1D texture sampling bug

mwyrzykowski opened the pull request:
#### 1b6e6ca8f0c2d663fbf331d0456529cfe44f951a
<pre>
WebGPU 1D texture sampling bug
<a href="https://bugs.webkit.org/show_bug.cgi?id=294223">https://bugs.webkit.org/show_bug.cgi?id=294223</a>
<a href="https://rdar.apple.com/152880664">rdar://152880664</a>

Reviewed by Tadeu Zagallo.

The docs for copyFromBuffer:...toTexture:...
do not allow for passing 0 to sourceBytesPerRow:
like they do for replaceRegion: calls. While it
seemed to work and is not detected by the Metal
validation layer, this appears to be a bug in WebKit
relying on undocumented behavior as it specifically
does not work in certain cases (sub buffer copies
from offsets not aligned to 64 bytes).

Test: api,operation,command_buffer,image_copy:* continues to pass, which
is the reason this zero was added.

Expectations are updated, all passing with this change.

* Source/WebGPU/WebGPU/Queue.mm:
(WebGPU::Queue::writeTexture):
(WebGPU::Queue::newTemporaryBufferWithBytes):

* LayoutTests/http/tests/webgpu/webgpu/api/operation/command_buffer/image_copy-expected.txt:
All sub-cases pass.

Canonical link: <a href="https://commits.webkit.org/296260@main">https://commits.webkit.org/296260@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2000a9e94a6817adeebe30517cfab5f54ebbb1b4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108005 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27670 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18089 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113216 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/58525 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28365 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36219 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/81990 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/110953 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22484 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97308 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62426 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/21896 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15441 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/57963 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/91837 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15507 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/116343 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35077 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/25821 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91019 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/35453 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93586 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/90813 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23134 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35716 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13473 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/30886 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/34975 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/40529 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34719 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38078 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36379 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

tadeuzagallo reviewed (approved):
Is this easy to test with the information from bugzilla? If it's not practical to test, it might be worth mentioning in the commit message how it was validated.

mwyrzykowski committed:
WebGPU 1D texture sampling bug
https://bugs.webkit.org/show_bug.cgi?id=294223
rdar://152880664

Reviewed by Tadeu Zagallo.

The docs for copyFromBuffer:...toTexture:...
do not allow for passing 0 to sourceBytesPerRow:
like they do for replaceRegion: calls. While it
seemed to work and is not detected by the Metal
validation layer, this appears to be a bug in WebKit
relying on undocumented behavior as it specifically
does not work in certain cases (sub buffer copies
from offsets not aligned to 64 bytes).

Test: api,operation,command_buffer,image_copy:* continues to pass, which
is the reason this zero was added.

Expectations are updated, all passing with this change.

* Source/WebGPU/WebGPU/Queue.mm:
(WebGPU::Queue::writeTexture):
(WebGPU::Queue::newTemporaryBufferWithBytes):

* LayoutTests/http/tests/webgpu/webgpu/api/operation/command_buffer/image_copy-expected.txt:
All sub-cases pass.

Canonical link: https://commits.webkit.org/296260@main
